Rubberduck.VBEEditor/ComManagement/WeakComSafe.cs

```diff
@@ -3,22 +3,26 @@
 using System.Runtime.CompilerServices;
 using Rubberduck.VBEditor.SafeComWrappers.Abstract;
 
+#if DEBUG
+using System.Linq;
+using System.Runtime.InteropServices;
+#endif
+
 namespace Rubberduck.VBEditor.ComManagement
 {
     public class WeakComSafe : IComSafe
     {
         //We use weak references to allow the GC to reclaim RCWs earlier if possible.
-        private readonly ConcurrentDictionary<int, WeakReference<ISafeComWrapper>> _comWrapperCache = new ConcurrentDictionary<int, WeakReference<ISafeComWrapper>>();
-
+        private readonly ConcurrentDictionary<int, (DateTime insertTime, WeakReference<ISafeComWrapper> weakRef)> _comWrapperCache = new ConcurrentDictionary<int, (DateTime, WeakReference<ISafeComWrapper>)>();
 
         public void Add(ISafeComWrapper comWrapper)
         {
             if (comWrapper != null)
             {
                 _comWrapperCache.AddOrUpdate(
                     GetComWrapperObjectHashCode(comWrapper), 
-                    key => new WeakReference<ISafeComWrapper>(comWrapper), 
-                    (key, value) => new WeakReference<ISafeComWrapper>(comWrapper));
+                    key => (DateTime.UtcNow, new WeakReference<ISafeComWrapper>(comWrapper)), 
+                    (key, value) => (DateTime.UtcNow, new WeakReference<ISafeComWrapper>(comWrapper)));
```

Comment:
I was a bit too hasty and didn't read properly - the updateFactory parameter should be doing `value.insertTime`, not `DateTime.UtcNow` in order to avoid changing the insertion time. But I am wondering why is the `WeakReference` new'd up here without any checks to the previous reference. Should we be disposing the old `WeakReference` or something?

Reply:
You really should not dispose the COM wrapper to which the weak reference points. With all likelihood it is the wrapper passed to you. (They have the same object hash.) In the very unlikely case that there you got a new wrapper with exactly the same hash as one allready in the dictionary, I prefer to safe the new one.

Reply:
That makes sense but what about the case where the original `WeakReference` is now wrapping a null reference? My worry is that because it points to a SCW and if the SCW in turn wraps a null reference, the `WeakReference` would still think it's alive even though it's just an empty shell at this point. In that case, we need to dispose... right?

Reply:
Just to summarize the discussion - `Debug.Assert` will be added to the `updateValueFactory` primarily to verify whether this get called as under normal circumstances, all operations should be additions or deletions, not updates. There might be rare update due to hash collisions. We will investigate if this is a contributing factor or not first before any further changes beside correcting the `DateTime.UtcNow` => `value.insertTime` is needed.